### ISSUE TITLE
Questions about Workflow Context

### ISSUE THREAD

**Issue by agross** [elsa-workflows/elsa-core#1483]
Hello,

I'm evaluating Elsa Workflows for a new project. Right now I'm in the process of testing Workflow Contexts (for my domain data) and attempted to rebuild [the guide from the docs](https://elsa-workflows.github.io/elsa-core/docs/next/concepts/concepts-workflow-context).

I have a couple of questions that I hope can be resolved with this issue:

1. The guide states that the HTTP request should be received with GET, but sending data implies POST (perhaps just a documentation issue).
1. I tried using the same database for Elsa's and the domain data, but migrations [do not seem to work](https://github.com/elsa-workflows/elsa-core/issues/1453#issuecomment-913622515) when two domains are mixed in a single DB. Is this to be expected?
1. If the above is true, I will need to separate my domain data from Elsa's data. How does Elsa handle consistency with two transactional stores (two databases) without introducing 2-phase commits?
1. Is it possible to get a list of all HTTP URLs that are handled by Elsa workflows using HTTP Endpoints (e.g. for debugging)?
1. During development and workflow design I've seen several cases where the initial HTTP Endpoint becomes kind of unusable after some requests:

    ```sh
    $ curl --location --request POST 'http://localhost:5000/workflows/testme' --header 'Content-Type: application/json' --data-raw '{"Title": "Creating Workflow Driven Apps with Elsa"}' | jq
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100   483    0   416  100    67   1337    215 --:--:-- --:--:-- --:--:--  1553
    {
      "errorMessage": "The call is ambiguous and matches multiple workflows.",
      "workflows": [
        {
          "WorkflowInstanceId": "788432e18a7e4a099c4fa4ace46b87c7",
          "ActivityId": "db48feb7-fb08-482b-b33c-9c706153b163"
        },
        {
          "WorkflowInstanceId": "b9304b4cc0684dd7a9ad45afbb862a9e",
          "ActivityId": "db48feb7-fb08-482b-b33c-9c706153b163"
        },
        {
          "WorkflowInstanceId": "572a1c023ccf40d6a51f7c3107231b32",
          "ActivityId": "db48feb7-fb08-482b-b33c-9c706153b163"
        }
      ]
    }
    ```

    In this example `workflows/testme` will not longer work. I needed to change the initial handler to something else to get meaningful results. It can be triggered quite easily, e.g. by entering something that cannot be evaluated in an If/Else element.

1. The [workflow context guide](https://elsa-workflows.github.io/elsa-core/docs/next/concepts/concepts-workflow-context) uses a kind of magic `workflowContext` variable for the data that has initially be `POST`ed to the workflow's HTTP Endpoint. It also states that the HTTP Endpoint should be named `HttpEndpoint1`, but later no longer refers to it. I find that confusing.

Thank you for taking the time to answer my noob questions.

Alex

**Comment by sfmskywalker** [comment]
Hi @agross,

1. You're absolutely right, that is a mistake. I just pushed a correction. Thanks for catching!
2. There should be no problem running migrations of different contexts mixed in a single DB. In fact, I have been doing this for a customer with no issue. If you send me a repro I can try and investigate. Unless you are deriving a new DB context based on ElsaContext, in which case you are responsible for generating AND running your own migrations.
3. Elsa relies on the underlying persistence provider (presumably EF Core in your case), so it does not know anything about third party domains. However, we could open up the persistence API for at least EF Core a little bit such that you can provide your own DbConnection and use a single transaction (since you would be using a single DB) as described [here](https://docs.microsoft.com/en-us/ef/core/saving/transactions#cross-context-transaction). If this is something you need then feel free to create an issue for this (and perhaps even a PR 😀)
4. It's certainly possible but not (yet) out of the box. I want to offer a way for Swashbuckle to include HTTP Endpoint activities when generating the Open API specification as described in #1499 created just now. If you want to get a list right now you could query the Bookmarks table and select all records where `ModelType` is `HttpEndpoint` and parse out the `Model` column into a `HttpEndpointBookmark` object, which gives you both the URL and the method.
5. I've seen this happening in the currently released packages (2.2.1) but was fixed in later packages (2.2.3 prerelease available from MyGet). If you are not using the latest prerelease then that might be the reason you are facing this bug.
6. That is indeed very confusing. I removed that bit, hopefully nobody else noticed 🙈 About the magic, that happens in the custom workflow context provider described earlier in the guide. It basically presumes that if the `workflowContext` is not set (i.e. is null), then the activity executing is the first one: the HttpEndpoint. Which will have received input in the form of an `HttpRequestModel`. This is provided by the HTTP Endpoint Middleware, which is what powers HTTP Endpoint activities. The mechanism is straightforward enough once you know: All HTTP requests flow through ASP.NET's middleware pipeline. One of these middleware is Elsa's HTTP Endpoint Middleware. It checks to see if the request path & method matches any HTTP Endpoint on any workflow. If a match is found, it executes that workflow and passes an HttpRequestModel object as input. Before the workflow executes, Elsa invokes the custom context provider, which takes the input, extracts a BlogPost out of it which then gets stored in the workflow context property of the workflow execution context.

I hope this clarifies things a little bit, but let me know if you have any follow-up questions.

**Comment by agross** [comment]
Hi @sfmskywalker,

Many thanks for your reply! I updated the Elsa libs to the preview versions and will let you know if the URL errors show up again.

I have a follow-up question regarding workflow design: Let's say the data for a workflow is managed externally, e.g. blog posts in a separate database. For the publishing workflow described in the workflow context guide, the initial request would not create the blog post, but rather load it from storage, perhaps setting some property denoting that the publishing workflow has started. The model that needs to be posted to the initial HTTP endpoint (e.g. `PublishingStarted(BlogPostId)`) would not necessarily need to match the actual `BlogPost` workflow context loaded from storage. When a new workflow is started, the `LoadWorkflowContext.WorkflowExecutionContext.IsFirstPass` flag seems to indicate that a new instance is started. The question is how to handle "not found" scenarios. Is this something that should be done in the workflow itself? Or am I heading in the wrong direction?

Another question that came up is related to specifying URLs, e.g. via docker. The docs mostly show a dashboard + server combination with `ELSA__SERVER__BASEURL` being defined as an environment variable. I understand that the server needs to know to URL to itself to e.g. generate links and the dashboard needs to know the public URL of the server to communicate to its API. I found `ELSA__SERVER__BASEADDRESS` and `ELSA__SERVER__BASEURL` in various examples. Which is used for which purpose? 

Thank you again for taking the time to answer my questions!

Alex
